### PR TITLE
[Bugfix][LoRA] Fix PEFT 0.18+ target_parameters LoRA loading for 3D MoE experts

### PR DESCRIPTION
**Pull request by darrenpeng-amazon**
# [Bugfix][LoRA] Fix PEFT 0.18+ target_parameters LoRA loading for 3D MoE experts

## Problem

PEFT 0.18+ can train a LoRA against a fused 3D MoE expert parameter via
`target_parameters` (e.g. `mlp.experts.gate_up_proj` / `down_proj`), reading it
`[E, dim1, dim2]` as `[E, in, out]`. vLLM's `add_lora_w13` / `add_lora_w2`
punica kernels use the opposite `nn.Linear` `[out, in]` convention, so per
expert `ΔW_vllm = ΔW_peft^T`. On stock vLLM, `_stack_moe_lora_weights` reshapes
with the wrong semantics and `set_lora` raises a shape mismatch that drops the
adapter (e.g. Qwen3.5-35B-A3B `down_proj`: `512 vs 2048`).

## Fix

- `peft_helper.py`: parse `target_parameters` from `adapter_config.json`.
- `lora_model.py`: propagate it onto `LoRAModel` (ctor, `clone()`,
  `from_lora_tensors`).
- `model_manager.py`: in `_stack_moe_lora_weights`, detect the PEFT 3D layout
  and swap `lora_A`/`lora_B` + transpose the trailing two dims **after** the
  EP-aware expert slicing.

Detection is required, not assumed: `create_dummy_lora_weights` builds dummies
in the native layout with the same flat shape, so an unconditional transpose
would corrupt them. Primary signal is `target_parameters` naming a fused expert
weight; fallback is `lora_a` trailing dim != `hidden_size`.

## Testing

```
python -m pytest -v \
  tests/lora/test_peft_helper.py::test_peft_helper_target_parameters_direct \
  tests/lora/test_lora_manager.py::test_target_parameters_indicates_3d_lora \
  tests/lora/test_lora_manager.py::test_from_lora_tensors_propagates_target_parameters \
  tests/lora/test_peft_helper.py::test_peft_helper_pass \
  tests/lora/test_lora_manager.py::test_from_lora_tensors
```

→ **10 passed** (the discriminator test is parametrized over 6 cases).

**End-to-end (Qwen3.5-35B-A3B expert LoRA).** The raw PEFT adapter crashes the
unpatched loader with the `512 vs 2048` error. With the fix, the directly-loaded
adapter is **bit-identical** (chosen-token logprobs, 32 tokens × 5 prompts) to
an independent offline-converted oracle that self-checks
`max|ΔW_vllm − ΔW_peft^T| = 0` over 320 experts, while both shift logits
measurably vs base — so the adapter takes effect and the in-loader transpose
matches the verified transform. Logprobs, not decoded text: for an experts-only
(8-of-256, rank-32) adapter, greedy text is a false-negative trap since the
adapter shifts logits without always flipping the argmax on short prompts.

## Latency impact

None expected. The transpose runs inside `_stack_moe_lora_weights` — the
adapter-load path, once per adapter (two `.transpose(-2, -1).contiguous()`
per fused expert weight). The per-forward hot path
(`add_lora_w13` / `add_lora_w2`) is not touched. 

## AI assistance disclosure

AI assistance was used to author this change. A human submitter has reviewed
every changed line and run the tests above.

### CODE REVIEW

**Comment by jeejeelee** [comment]
cc @linitra24

**Comment by linitra24** [comment]
@darrenpeng-amazon  Thanks for the PR. I believe this issue was already fixed upstream in https://github.com/huggingface/peft/pull/3165, released in v0.19.1.

I also tested a 3D MoE LoRA locally with PEFT 0.20.0 and could not reproduce the shape mismatch. The generated A/B tensors already match vLLM’s expected layout.

Since newer adapters also contain `target_parameters`, using it to trigger an unconditional A/B swap may corrupt valid adapters. Therefore, I don’t think this vLLM-side change is needed now.
